gk/bpf.c

```diff
@@ -16,13 +16,157 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#define GK_BPF_INTERNAL static
+#include "gatekeeper_flow_bpf.h"
+
 #include "gatekeeper_gk.h"
+#include "gatekeeper_main.h"
+
 #include "bpf.h"
 
 static const struct rte_bpf_xsym flow_handler_init_xsym[] = {
 };
 
+struct gk_bpf_pkt_frame {
+	uint64_t		password;
+	struct flow_entry	*fe;
+	struct rte_mbuf		*pkt;
+	struct gk_config	*gk_conf;
+	struct gk_bpf_pkt_ctx	ctx;
+};
+
+static const uint64_t pkt_password = 0xa2e329ba8b15af05;
+
+static struct gk_bpf_pkt_frame *
+pkt_ctx_to_frame(struct gk_bpf_pkt_ctx *ctx)
+{
+	struct gk_bpf_pkt_frame *frame;
+
+	if (unlikely(ctx == NULL))
+		return NULL;
+
+	frame = container_of(ctx, struct gk_bpf_pkt_frame, ctx);
+	if (unlikely(frame->password != pkt_password)) {
+		GK_LOG(WARNING, "%s(): password violation\n", __func__);
+		return NULL;
+	}
+
+	return frame;
+}
+
+static struct gk_bpf_cookie *
+pkt_ctx_to_cookie(struct gk_bpf_pkt_ctx *ctx)
+{
+	struct gk_bpf_pkt_frame *frame = pkt_ctx_to_frame(ctx);
+	if (unlikely(frame == NULL))
+		return NULL;
+	return &frame->fe->u.bpf.cookie;
+}
+
+static struct rte_mbuf *
+pkt_ctx_to_pkt(struct gk_bpf_pkt_ctx *ctx)
+{
+	struct gk_bpf_pkt_frame *frame = pkt_ctx_to_frame(ctx);
+	if (unlikely(frame == NULL))
+		return NULL;
+	return frame->pkt;
+}
+
+static int
+gk_bpf_encapsulate(struct gk_bpf_pkt_ctx *ctx, int priority)
+{
+	struct gk_bpf_pkt_frame *frame = pkt_ctx_to_frame(ctx);
+	if (unlikely(frame == NULL))
+		return -EINVAL;
+
+	if (unlikely(priority < 0 || priority > PRIORITY_MAX))
+		return -EINVAL;
+
+	return encapsulate(frame->pkt, priority, &frame->gk_conf->net->back,
+		&frame->fe->grantor_fib->u.grantor.gt_addr);
+}
+
 static const struct rte_bpf_xsym flow_handler_pkt_xsym[] = {
+	{
+		.name = "cycles_per_sec",
+		.type = RTE_BPF_XTYPE_VAR,
+		.var = {
+			.val = &cycles_per_sec,
+			.desc = {
+				.type = RTE_BPF_ARG_RAW,
+				.size = sizeof(cycles_per_sec),
+			},
+		},
+	},
+	{
+		.name = "cycles_per_ms",
+		.type = RTE_BPF_XTYPE_VAR,
+		.var = {
+			.val = &cycles_per_ms,
+			.desc = {
+				.type = RTE_BPF_ARG_RAW,
+				.size = sizeof(cycles_per_ms),
+			},
+		},
+	},
+	{
+		.name = "pkt_ctx_to_cookie",
+		.type = RTE_BPF_XTYPE_FUNC,
+		.func = {
+			.val = (void *)pkt_ctx_to_cookie,
+			.nb_args = 1,
+			.args = {
+				[0] = {
+					.type = RTE_BPF_ARG_PTR,
+					.size = sizeof(struct gk_bpf_pkt_ctx),
+				},
+			},
+			.ret = {
+				.type = RTE_BPF_ARG_PTR,
+				.size = sizeof(struct gk_bpf_cookie),
+			},
+		},
+	},
+	{
+		.name = "pkt_ctx_to_pkt",
+		.type = RTE_BPF_XTYPE_FUNC,
+		.func = {
+			.val = (void *)pkt_ctx_to_pkt,
+			.nb_args = 1,
+			.args = {
+				[0] = {
+					.type = RTE_BPF_ARG_PTR,
+					.size = sizeof(struct gk_bpf_pkt_ctx),
+				},
+			},
+			.ret = {
+				.type = RTE_BPF_ARG_PTR_MBUF,
+				.size = sizeof(struct rte_mbuf),
```

Comment:
For mbuf ptr type, it should be [the max size of rte_mbuf data buffer](https://github.com/cjdoucette/dpdk/blob/a610503483894a07abd163b241aebb9119b34e99/lib/librte_bpf/rte_bpf.h#L48-L49).

Reply:
That comment is a little confusing since it's referring to field `buf_size` not to field `size`. Anyway, I have forgotten to set the field `buf_size` and I'm going to do so now. For an example of using `RTE_BPF_ARG_PTR_MBUF`, see `bpf_parse_flags()` in file `app/test-pmd/bpf_cmd.c`.